core/src/main/scala/com/mesosphere/usi/core/SchedulerLogic.scala

```diff
@@ -0,0 +1,251 @@
+package com.mesosphere.usi.core
+
+import java.time.Instant
+
+import com.mesosphere.usi.core.models._
+import com.mesosphere.usi.models._
+
+object SchedulerLogic {
+  private def assertValidTransition(oldSpec: PodSpec, newSpec: PodSpec): Unit = {
+    if ((oldSpec.goal == Goal.Terminal) && (newSpec.goal == Goal.Running))
+      throw new IllegalStateException(s"Illegal state transition Terminal to Running for podId ${oldSpec.id}")
+  }
+
+  def taskIdsFor(pod: PodSpec): Seq[TaskId] = {
+    // ignoring momentarily that these might be different
+    Seq(pod.id.value)
+  }
+
+  def handlePodSpecUpdate(podSpecs: Map[PodId, PodSpec], podRecords: Map[PodId, PodRecord])
+                         (podId: PodId, newState: Option[PodSpec], responseBuilder: ResponseBuilder): ResponseBuilder = {
+    var b = responseBuilder
+
+    newState match {
+      case None =>
+        b = b.withChangeMessage(PodStatusChange(podId, None))
+        if (podRecords.contains(podId)) {
+          b = b.withChangeMessage(PodRecordChange(podId, None))
+        }
+
+      case Some(podSpec) =>
+
+        podSpecs.get(podId).foreach(assertValidTransition(_, podSpec))
+
+        if (podSpec.goal == Goal.Terminal) {
+          taskIdsFor(podSpec).foreach { taskId =>
```

Comment:
I hope we're gonna extract the killing logic out of here later. E.g. I don't think we should kill every time you receive update. Do we have JIRA to implement proper killing?

Reply:
Yes, that is the intention. I can add a comment make that more clear. No specific JIRA, but, then again, were not even doing proper test killing anyways. As an aside, is there really a downside to sending a kill on every terminal update? It's kind of an idempotent thing.